navigation/src/androidMain/kotlin/com/freeletics/khonshu/navigation/NavHost.kt

```diff
@@ -203,36 +202,45 @@ private fun systemBackHandling(snapshot: StackSnapshot, navigator: HostNavigator
     val backProgress = remember(snapshot) {
         Animatable(0f, visibilityThreshold = VISIBILITY_THRESHOLD)
     }
-    PredictiveBackHandler(enabled = snapshot.canNavigateBack) { progressFlow ->
-        var finalValue = 0f
+    PredictiveBackHandler(
+        enabled = snapshot.canNavigateBack,
+        extraCallback = navigator.onBackPressedCallback,
+    ) { progressFlow ->
         try {
             progressFlow.collect { backEvent ->
                 backProgress.snapTo(backEvent.progress)
             }
-            finalValue = 1f
-            backProgress.animateTo(1f)
-            navigator.navigateBack()
+            backProgress.tryAnimateTo(1f)
+            navigator.tryNavigateBack()
         } catch (e: CancellationException) {
-            backProgress.animateTo(0f)
-        } finally {
-            // make sure that the animation is not stuck at intermediate value in case animateTo gets cancelled
-            backProgress.snapTo(finalValue)
+            backProgress.tryAnimateTo(0f)
         }
     }
 
-    // needs to be called after PredictiveBackHandler because the navigator has precedence
-    val backPressedDispatcher = requireNotNull(LocalOnBackPressedDispatcherOwner.current) {
-        "No OnBackPressedDispatcher available"
-    }
-    DisposableEffect(backPressedDispatcher) {
-        backPressedDispatcher.onBackPressedDispatcher.addCallback(navigator.onBackPressedCallback)
+    return backProgress.asState()
+}
 
-        onDispose {
-            navigator.onBackPressedCallback.remove()
+private fun HostNavigator.tryNavigateBack() {
+    try {
+        navigateBack()
+    } catch (e: IllegalStateException) {
+        // The exception is thrown when navigateBack is called while the backstack is at the root. If the
+        // system back is triggered twice very quickly after each other there is a short time window
+        // after the first where the OnBackPressedCallback is not yet updated and would then also handle the
+        // second. This suppresses the crash in that case.
+        if (snapshot.value.canNavigateBack) {
+            throw e
         }
     }
+}
 
-    return backProgress.asState()
+private suspend fun Animatable<Float, *>.tryAnimateTo(value: Float) {
+    try {
+        animateTo(value)
+    } finally {
+        // make sure that the animation is not stuck at intermediate value in case animateTo gets cancelled
+        snapTo(value)
+    }
 }
```

Comment:
This is better than the final value approach because even if `animateTo` is cancelled we should still call `navigateBack`.